apps/total-typescript/src/lib/bookmarks.ts

```diff
@@ -0,0 +1,158 @@
+import {z} from 'zod'
+import {prisma} from '@skillrecordings/database'
+
+export const BookmarkInputSchema = z.object({
+  userId: z.string(),
+  resourceId: z.string(),
+  type: z.string(),
+  fields: z.record(z.string().nullable()).optional(),
+})
+
+export const ReturnedBookmarkSchema = z.object({
+  id: z.string(),
+  resourceId: z.string(),
+  type: z.string(),
+  fields: z.record(z.string().nullable()).optional(),
+})
+
+export const ReturnedBookmarkForBookSchema = z.object({
+  id: z.string(),
+  resourceId: z.string(),
+  type: z.string(),
+  fields: z.object({
+    chapterSlug: z.string(),
+    chapterTitle: z.string(),
+    resourceTitle: z.string(),
+  }),
+})
+
+export type Bookmark = z.infer<typeof ReturnedBookmarkSchema>
+export type BookmarkForBook = z.infer<typeof ReturnedBookmarkForBookSchema>
+export type BookmarkInput = z.infer<typeof BookmarkInputSchema>
+
+export async function getBookmarksForUser(userId: string) {
+  try {
+    const bookmarks = await prisma.bookmark.findMany({
+      select: {
+        id: true,
+        resourceId: true,
+        type: true,
+        fields: true,
+      },
+      where: {
+        userId,
+      },
+    })
+    return bookmarks
+  } catch (error) {
+    console.error('Error getting bookmarks for user', error)
+    throw new Error('Error getting bookmarks for user')
+  }
+}
+
+export async function getBookmarkByResourceId({
+  userId,
+  resourceId,
+}: {
+  userId: string
+  resourceId: string
+}) {
+  try {
+    const bookmark = await prisma.bookmark.findFirst({
+      select: {
+        id: true,
+        resourceId: true,
+        type: true,
+        fields: true,
+      },
+      where: {
+        userId,
+        resourceId,
+      },
+    })
+    return bookmark
+  } catch (error) {
+    console.error('Error getting bookmark', error)
+    throw new Error('Error getting bookmark')
+  }
+}
+
+export async function getLastestBookmarkedResources({
+  userId,
+  type,
+}: {
+  userId: string
+  type: string
+}) {
+  try {
+    const latestBookmarkResources = await prisma.bookmark.findMany({
+      select: {
+        id: true,
+        resourceId: true,
+        type: true,
+        fields: true,
+      },
+      where: {
+        type,
+        userId,
+      },
+      orderBy: {
+        createdAt: 'desc',
+      },
+    })
+    return latestBookmarkResources
```

Comment:
we can't know at the database level if the latest bookmark is for the current module so we return all of the users bookmarks filtered on the type and let the frontend do further filtering to grab the latest for the relevant module

Reply:
you probably can with sql but this is OK too